### PR TITLE
[FastPR][Core] Operations addition

### PR DESCRIPTION
**Pull request by rubenzorrilla**
**📝 Description**
This PR adds the base class for `Operations`. This will be used later on in the multistage.

@KratosMultiphysics/technical-committee I decided to separate it from the registry / multistage stuff to make the final review easier.

EDIT: @KratosMultiphysics/technical-committee note that I already added the registry macro at the bottom of `operation.h`.

### CODE REVIEW

**Review by philbucher — DISMISSED** [review]

File: kratos/operations/operation.h [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+    {                                                                                                           \
+        std::string all_path = std::string("Operations.All.") + operation_name;                                 \
+        if (!Registry::HasItem(all_path)) {                                                                     \
+            auto& r_operation_item = Registry::AddItem<RegistryItem>(all_path);                                 \
+            r_operation_item.AddItem<RegistryValueItem<Operation>>("Prototype", operation_prototype);           \
+        } else {                                                                                                \
+            KRATOS_ERROR << "Operation '" << operation_name << "' is already registered." << std::endl;         \
+        }                                                                                                       \
+        std::string module_path = std::string("Operations.") + module_name + std::string(".") + operation_name; \
+        if (!Registry::HasItem(module_path)) {                                                                  \
```
Comment (pooyan-dadvand) [inline kratos/operations/operation.h:166]: Shall we do this check? or we should leave it to registry to send an error?

File: kratos/operations/operation.h [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+///@{
+
+#ifndef KRATOS_REGISTER_OPERATION
+#define KRATOS_REGISTER_OPERATION(                                                                              \
+    module_name,                                                                                                \
+    operation_name,                                                                                             \
+    operation_prototype)                                                                                        \
+    {                                                                                                           \
+        std::string all_path = std::string("Operations.All.") + operation_name;                                 \
+        if (!Registry::HasItem(all_path)) {                                                                     \
```
Comment (pooyan-dadvand) [inline kratos/operations/operation.h:159]: I understand that you are checking to send a custom error. Isn't it?

If this is the case, we may need to create (not in this PR) a AddItem(path, error_lambda) to make this more straightforward

**Review by pooyan-dadvand — COMMENTED** [review]
I don't revoke the approval but there are some minor comments

File: kratos/operations/operation.h (hunk 1 above)
Comment (rubenzorrilla) [reply kratos/operations/operation.h:166]: (see above answer)

File: kratos/operations/operation.h (hunk 2 above)
Comment (rubenzorrilla) [reply kratos/operations/operation.h:159]: True. As it is right now we're checking twice, and the custom error isn't that custom... About the the `AddItem` with error thrown, it is not a bad idea.

Considering that we're about to start the PR with all the new registry stuff, and that this macro is almost identical to that of the `Process` that I'll add in there, I think we'd better discuss it in the upcoming one.

**Review by sunethwarna — APPROVED** [review]

**Comment by rubenzorrilla** [comment]
`ShallowWaterApplication` tests are failing. Strange as these have nothing to do with this PR. @miguelmaso FYI.

File: kratos/operations/operation.h (hunk 2 above)
Comment (pooyan-dadvand) [reply kratos/operations/operation.h:159]: Agree that we shall do it in a separate PR

File: kratos/operations/operation.h (hunk 1 above)
Comment (pooyan-dadvand) [reply kratos/operations/operation.h:166]: But here you are avoiding the error. Is it intentional?

**Comment by loumalouomega** [comment]
> `ShallowWaterApplication` tests are failing. Strange as these have nothing to do with this PR. @miguelmaso FYI.

Timeout, maybe some small tests must be moved to nightly

**Comment by miguelmaso** [comment]
> > `ShallowWaterApplication` tests are failing. Strange as these have nothing to do with this PR. @miguelmaso FYI.
> 
> Timeout, maybe some small tests must be moved to nightly

Does anyone has more information about the failing test? It seems to be a specific test (Boussinesq) with clang but I don't know how to reproduce it

**Comment by philbucher** [comment]
no clue
did you try to reproduce it locally?

**Comment by roigcarlo** [comment]
Trying in my machine but so far works...

**Pull request merged**